towhee/runtime/pipeline_manager.py

```diff
@@ -0,0 +1,109 @@
+
+
+class PipelineThread(threading.Thread):
+    """
+    Pipeline thread.
+
+    Args:
+        nodes(`Dict[str, NodeRepr]`): The pipeline nodes from DAGRepr.nodes.
+        edges(Dict[str, List]`: The pipeline edges from DAGRepr.edges.
+        operator_pool(`OperatorPool`): The operator pool.
+    """
+    def __init__(self, nodes: Dict[str, NodeRepr], edges: Dict[str, Any], operator_pool: 'OperatorPool'):
+        super().__init__()
+        self._nodes = nodes
+        self._edges = edges
+        self._operator_pool = operator_pool
+        self._thread_pool = ThreadPoolExecutor()
+        self._node_runners = None
+        self._data_queues = dict((name, DataQueue(edge['data'])) for name, edge in self._edges.items())
+        self.setDaemon(True)
+
+    def initialize(self):
+        self._node_runners = []
+        for name in DAGRepr.get_top_sort(self._nodes):
+            in_queues = [self._data_queues[edge] for edge in self._nodes[name].in_edges]
+            out_queues = [self._data_queues[edge] for edge in self._nodes[name].out_edges]
+            node = create_node(self._nodes[name], self._operator_pool, in_queues, out_queues)
+            if not node.initialize():
+                raise RuntimeError(node.err_msg)
+            self._node_runners.append(node)
+
+    def result(self) -> any:
+        end_edge_num = self._nodes['_output'].out_edges[0]
+        res = self._data_queues[end_edge_num]
+        if res.size != 0:
+            return res.get()
+        else:
+            # run failed, raise an exception
+            for node in self._node_runners:
+                if node.status == NodeStatus.FAILED:
+                    raise RuntimeError(node.err_msg)
+            engine_log.warning('The pipeline runs successfully, but no data return')
+            return None
+
+    def __call__(self, *inputs):
+        self.initialize()
+        self._data_queues[0].put(*inputs)
+        self._data_queues[0].seal()
+        for node in self._node_runners:
+            self._thread_pool.submit(node.process)
+        return self.result()
+
+    def release(self):
+        del self._data_queues
+        del self._node_runners
+
+
+class PipelineManager:
+    """
+    Manage the pipeline and runs it as a single instance.
+
+
+    Args:
+        dag_repr(`Dict`): A DAGRepr or Dictionary from the user pipeline.
+    """
+
+    def __init__(self, dag_repr: Dict):
+        self._dag_repr = DAGRepr.from_dict(dag_repr)
+        self._operator_pool = OperatorPool()
+
+    def preload(self):
+        """
+        Initialize the nodes.
+        """
+        pipeline_context = PipelineThread(self._dag_repr.nodes, self._dag_repr.edges, self._operator_pool)
+        pipeline_context.initialize()
+
+    def __call__(self, *inputs):
+        """
+        Output with ordering matching the input `DataQueue`.
+        """
+        pipeline_context = PipelineThread(self._dag_repr.nodes, self._dag_repr.edges, self._operator_pool)
```

Comment:
It is inefficient to create threads for each `__call___`

Reply:
If we run with one thread from PipelinManager, it will always create the same pipeline_context object, and how to release the memory?